src/v/cluster/partition.cc

```diff
@@ -1213,19 +1213,52 @@ partition::get_cloud_storage_manifest_view() {
 ss::future<result<model::offset, std::error_code>>
 partition::sync_kafka_start_offset_override(
   model::timeout_clock::duration timeout) {
-    if (_log_eviction_stm && !is_read_replica_mode_enabled()) {
+    if (is_read_replica_mode_enabled()) {
+        auto term = _raft->term();
+        if (!co_await _archival_meta_stm->sync(timeout)) {
+            if (term != _raft->term()) {
+                co_return errc::not_leader;
+            } else {
+                co_return errc::timeout;
+            }
+        }
+        auto start_kafka_offset
+          = _archival_meta_stm->manifest().get_start_kafka_offset_override();
+
+        co_return kafka::offset_cast(start_kafka_offset);
+    }
+
+    if (_log_eviction_stm) {
         auto offset_res = co_await _log_eviction_stm
                             ->sync_kafka_start_offset_override(timeout);
         if (offset_res.has_failure()) {
             co_return offset_res.as_failure();
         }
         if (offset_res.value() != kafka::offset{}) {
-            co_return offset_res.value();
+            co_return kafka::offset_cast(offset_res.value());
```

Comment:
Should we have set _cached_kafka_start_override here?

Reply:
We could, it wouldn't change anything though since `sync_kafka_start_offset_override` is already returning an internally cached offset. The `_cached_kafka_start_override` is really only being used to indicate that we've sync'd the archival stm and have the most recent kafka start offset override from cloud storage. I could potentially even just make it a bool. `has_synced_archival` or something.

Reply:
I see, so the fundamental shift here is that the log eviction stm will always know the kafka offset, unless there's been a restart, in which case we can fall back on archival. And subsequent delete-records after the restart will be caught by the log eviction stm.

Could you incorporate that into the comments here and in log_eviction_stm? You touch on it a little bit, but I think we should be more explicit with when we expect log_eviction_stm to fall back on archival

Reply:
I've update the comments in the log eviction stm and the partition class to better indicate what's going on.